design/open/staging-compaction.md

```diff
@@ -0,0 +1,172 @@
+
+
+## Non Goals
+
+- Improve the performance of the KV store in general.
+- Simplify the uncommitted area model in lakeFS. While we'll try to keep the 
+  changes minimal, we might need to add some complexity to the model to 
+  achieve the goals.
+
+
+## Proposed Design
+
+We propose to add a compaction mechanism to the staging area. We'll explain 
+how to decide when to compact (Sensor), how to compact (Compactor), 
+data read/write flows and commit flows after compaction.
+
+```go
+type Branch struct {
+	CommitID     CommitID
+	StagingToken StagingToken
+	// SealedTokens - Staging tokens are appended to the front, this allows building the diff iterator easily
+	SealedTokens []StagingToken
+	
+	CompactedMetaRange MetaRangeID
+	CompactionStartTime time.Time
+}
+```
+
+### Sensor
+
+The Sensor is responsible for deciding when to compact the staging area.
+The Sensor will be linked to the Graveler and will collect 
+information on writes to the staging area. It will decide when to compact a 
+certain branch based on the number of deleted entries to its staging area, 
+taking into accounts operations like commits/resets etc. It can also decide 
+based on the number of sealed tokens, although not necessarily a priority 
+for the first version. We can probably avoid the need to query the kv-store 
+to retrieve that information (except for the service startups) by caching 
+the information in the Sensor.
+
+Notice there's a single Sensor for each lakeFS. While the followup sections 
+describe why concurrent compactions don't harm consistency, they may be very 
+inefficient. Therefore, a Sensor deciding to compact will only do so if the 
+branch's `CompactionStartTime` is not within the last x<TBD> minutes using 
+`SetIf` to minimize collisions (although not avoiding them completely due to 
+clock skews). 
+
+### Compactor
+
+Upon deciding to compact, the Sensor will trigger the Compactor. The 
+Compactor will perform an operation very similar to Commit. Before starting 
+to compact, it will atomically:
+
+1. Push a new `StagingToken` to the branch.
+1. Add the old `StagingToken` to the `SealedTokens` list.
+1. Set the `CompactionStartTime` to the current time.
+
+The Compactor will now read the branch's `SealedTokens` in order and apply 
+them on either the `CompactedMetaRange` if it exists, or on the branch's 
+HEAD CommitID MetaRange to create a newMetaRange. The Compactor will then 
+atomically update the branch entity:
+
+1. Set the `CompactedMetaRangeID` to the new MetaRangeID.
+1. Remove the compacted `SealedTokens` from the branch.
+
+The Compactor should fail the operation if the sealed tokens have changed 
+since the compaction started (SetIf). Although the algorithm can tolerate 
+additional sealed tokens being added during the compaction, it's better to 
+avoid competing with concurrent commits. Commits have the same benefits as 
+compactions, but they are proactively triggered by the user (and might fail 
+if compaction succeeds).
```

Comment:
This the extra cool property that at most one compaction will start concurrently. :-)